### DISCUSSION TITLE
SSR session access from custom app (_app)

### DISCUSSION THREAD

**Discussion by giulianok** [nextauthjs/next-auth#5499] [Help]
### Question 💬

I'm currently trying to access to the `session` via server since I have a layout setup at the app level through custom _app (see )

Here's an example:

```tsx
const MyApp: AppType = ({
  Component,
  pageProps: { session, ...pageProps },
}) => {
  return (
    <SessionProvider session={session} refetchOnWindowFocus={false}>
      <Layout>
        <Component {...pageProps} />
      </Layout>
    </SessionProvider>
  )
}
```

Right now, `session` is `undefined` on the server, not sure why it is "required" to be passed to `SessionProvider` since it doesn't really change if I do or don't pass it.

`unstable_getServerSession` could be a solution but I cannot use it in `_app` since a custom App does not support `getServerSideProps` and `getStaticProps`, it only supports `getInitialProps` where `unstable_getServerSession` is not supported.

### How to reproduce ☕️

```tsx
const MyApp: AppType = ({
  Component,
  pageProps: { session, ...pageProps },
}) => {
  return (
    <SessionProvider session={session} refetchOnWindowFocus={false}>
      <Layout>
        <Component {...pageProps} />
      </Layout>
    </SessionProvider>
  )
}
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ThangHuuVu** [comment]
It is not required to pass `session` to `SessionProvider`, doing so will help avoid checking on the session twice on the client.
From the doc of `SessionProvider` https://next-auth.js.org/getting-started/client#sessionprovider

> If you pass the session page prop to the <SessionProvider> – as in the example above – you can avoid checking the session twice on pages that support both server and client-side rendering.

Code here: https://github.com/nextauthjs/next-auth/blob/0a4b99de3b7609b3994086ccddfdc67b147557d7/packages/next-auth/src/react/index.tsx#L319

To get the session on server side, can just use `unstable_getServerSession` on any pages other than `_app.tsx` 👀

**Reply by clement-buchart** [reply]
Hello, 

On the page you mentioned, it also says : 

> If every one of your pages needs to be protected, you can do this in getInitialProps in _app, otherwise you can do it on a page-by-page basis. 

Yet as @giulianok mentionned unstable_getServerSession doesn't seem to work when used in getInitialProps in _app. 

Is there something I am missing ?

**Reply by giulianok** [reply]
Correct but the problem is that 1) since we're using the Layout approach, we have a component called `Header` that checks if the user is signed in and renders its name, otherwise it renders a login button, this doesn't work in SSR since `unstable_getServerSession` cannot be used in `_app` and the Layout is consumed in the `_app` as https://nextjs.org/docs/basic-features/layouts, and 2) the `page-by-page basis` approach will lead to a ton of code duplication

**Reply by clement-buchart** [reply]
Yes I think we agree. 
I have a similar requirement on my app. 

The thing I would like to understand is why is the documentation specifically stating "you can do this in getInitialProps in _app" in spite of it not working apparently for you and me.

**Reply by giulianok** [reply]
this is what I get when I run `unstable_getServerSession` in `getInitialProps` 

![Screen Shot 2022-10-13 at 1 57 56 PM](https://user-images.githubusercontent.com/8062245/195675935-615e0086-0846-4c3a-9831-bdbd608edd20.png)

**Reply by clement-buchart** [reply]
Yes, same thing.

I'm no expert on the matter but my best guess is that the static method `getInitialProps` does not benefit from the [next-code-elimination tool](https://next-code-elimination.vercel.app/) like the exported func `getServerSideProps` does (consistent with [next docs](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#when-does-getserversideprops-run) : `Note that you must export getServerSideProps as a standalone function — it will not work if you add getServerSideProps as a property of the page component.`), so v8 which is exclusively available on the server side crash the client side.

I'm not sure anything can be done unless Next offer a getServerSideProps equivalent for app.

For now, i created a function like 
```
export const withSession = async (
  otherProps: { props?: any },
  ctx: GetServerSidePropsContext
) => {
  return {
    props: {
      ...otherProps.props,
      session: await unstable_getServerSession(ctx.req, ctx.res, authOptions),
    },
  };
};
```

and in the page I do something like 
```
export const getServerSideProps: GetServerSideProps = async (
  ctx: GetServerSidePropsContext
) => {
  return withSession(
    {
      props: {
        foo: "bar",
      },
    },
    ctx
  );
};
```

Doesn't reduce code duplication by a lot in this case, but in my case I have other stuff I want to make globally available and that depend on session (namely a graphQL client, so I also have a `withGraphQL` wrapper)

**Comment by giulianok** [comment]
Gotcha, so that means that the session will always be fetched if the SessionProvider is consumed from the `_app.tsx` since `hasInitialSession` is `undefined` (https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/react/index.tsx#L312). The only way to avoid it is by getting the session with `unstable_getServerSession` via `getServerSideProps` which is only possible at a page le🤔vel.

About `getServerSideProps`, the reason I'm looking for a way to get the session from the `_app.tsx` is because I have a Layout (https://nextjs.org/docs/basic-features/layouts) in my `_app.tsx`. 

Based on what I mentioned before, I would need to get the session through `getServerSideProps` and render the `SessionProvider` passing the session, on every single page, which leads to a lot of repeatable code per page
